rclcpp/test/rclcpp/test_publisher.cpp

```diff
@@ -259,3 +264,178 @@ TEST_F(TestPublisher, basic_getters) {
+  // Failure in rcl_publisher_fini should just log error
+  EXPECT_THROW(
+    node->create_publisher<test_msgs::msg::Empty>("topic", 10).reset(),
+    rclcpp::exceptions::RCLError);
+}
+
+TEST_F(TestPublisher, rcl_publisher_get_rmw_handle_error) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return("lib:rclcpp", rcl_publisher_get_rmw_handle, nullptr);
+  RCLCPP_EXPECT_THROW_EQ(
+    node->create_publisher<test_msgs::msg::Empty>("topic", 10),
+    std::runtime_error("failed to get rmw handle: error not set"));
+}
+
+TEST_F(TestPublisher, rcl_publisher_get_gid_for_publisher_error) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rmw_get_gid_for_publisher, RMW_RET_ERROR);
+  RCLCPP_EXPECT_THROW_EQ(
+    node->create_publisher<test_msgs::msg::Empty>("topic", 10),
+    std::runtime_error("failed to get publisher gid: error not set"));
+}
+
+TEST_F(TestPublisher, rcl_publisher_fini_error) {
+  initialize();
+  auto mock = mocking_utils::inject_on_return("lib:rclcpp", rcl_publisher_fini, RCL_RET_ERROR);
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  ASSERT_EQ(1, publisher.use_count());
+  EXPECT_NO_THROW(publisher.reset());
+}
+
+TEST_F(TestPublisher, rcl_publisher_get_options_error) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return("lib:rclcpp", rcl_publisher_get_options, nullptr);
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  RCLCPP_EXPECT_THROW_EQ(
+    publisher->get_queue_size(),
+    std::runtime_error("failed to get publisher options: error not set"));
+}
+
+TEST_F(TestPublisher, rcl_publisher_get_subscription_count_publisher_invalid) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_publisher_get_subscription_count, RCL_RET_PUBLISHER_INVALID);
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  EXPECT_THROW(
+    publisher->get_subscription_count(),
+    rclcpp::exceptions::RCLError);
+}
+
+TEST_F(TestPublisher, rcl_publisher_get_actual_qos_error) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rcl_publisher_get_actual_qos, nullptr);
+  auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  RCLCPP_EXPECT_THROW_EQ(
+    publisher->get_actual_qos(),
+    std::runtime_error("failed to get qos settings: error not set"));
+}
+
+TEST_F(TestPublisher, publishers_equal_rmw_compare_gids_error) {
+  initialize();
+  auto mock = mocking_utils::patch_and_return(
+    "lib:rclcpp", rmw_compare_gids_equal, RMW_RET_ERROR);
+  const auto publisher = node->create_publisher<test_msgs::msg::Empty>("topic", 10);
+  const rmw_gid_t * gid = nullptr;
+  auto throwing_fn = [publisher, gid]()
+    {
+      // The == operator is expected to throw here, but this lambda avoids unused result warning
+      return (*publisher.get() == gid) ? true : false;
```

Comment:
@brawner uh, can a publisher be compared to an `rmw_gid_t`? That's unusual.

Reply:
This is coverage this operator overload here:

https://github.com/ros2/rclcpp/blob/2531787550084307a5c8144a68c0195fcdd6442a/rclcpp/include/rclcpp/publisher_base.hpp#L176-L184